micropip/manager.py

```diff
@@ -0,0 +1,69 @@
+from typing import Any, List
+
+from micropip import package_index
+
+
+class Manager:
+    """
+    Manager provides an extensible interface for customizing micropip's behavior.
+
+    Each Manager instance holds its own local state that is
+    independent of other instances, including the global state.
```

Comment:
> including the global state.

Well it's not really global then. It's a bit unclear how this works since the result of installing stuff with micropip is generally files placed in the file system. I guess different `Manager` instances can have different install directories?

Reply:
I meant to phrase it in a way that the global manager instance(representing the 'global state' for the exposed commands) is also included in the statement, but maybe it isn't necessary to specify it as such.

I would guess that the 'right way' to keep things independent would be to also have different install directories, but I'm not sure how nicely that would play out with sitepackages, or perhaps a mechanism to prevent managers having duplicate install directories. Could you provide your opinion on this? cc: @ryanking13

Reply:
> I guess different Manager instances can have different install directories?

Ideally, we may try to have each instance behave completely independently as if each instance becomes a virtual environment. For example, by combining micropip and subinterpreter, we may be able to support different interpreters to install and run different sets of packages.

However, I think most of the time, people will end up using one instance, and we would recommend doing so, too.

Reply:
This is super useful for testing purposes but probably not as useful for real world stuff. How does `importlib` manage this? It seems to me that it depends on `sys.path` and `sys.meta_path` for everything so it can't handle introspecting multiple environments without monkeypatching these. Presumably there are pypi packages that can inspect virtual environments that aren't active.

Reply:
I know `importlib.metadata.distributions()` takes an optional `path` argument, which defaults to `sys.path`, allowing a caller to inspect metadata for a set of paths. That could be pointed at a `site-packages` for a virtualenv to get the metadata for those packages, _but_ it won't honor other paths that might get added by the site config (such as a `venv --system-site-packages` or a `.pth` file in the site-packages that might extend the path). In my experience, the only technique I've seen for inspecting the true packaging environment for an interpreter is to launch that interpreter (in a subprocess) and then inspect that interpreter's environment.

That said, I know recent versions of `tox` have been doing some innovative things around detecting installed packages (and diffing them against an install spec), so there may be some good ideas in that codebase.

Reply:
> This is super useful for testing purposes but probably not as useful for real world stuff.

My main goal of making micropip modular is to allow us (or others) to customize some of its behavior. It's nice to have each instance behave independently, but it's not really something I'm very interested in right now (of course we can benefit from it in testing purpose).